streams/src/main/java/org/apache/kafka/streams/kstream/internals/GroupedStreamAggregateBuilder.java

```diff
@@ -74,14 +75,18 @@
 
         final String aggFunctionName = builder.newProcessorName(functionName);
 
-        final OptimizableRepartitionNode.OptimizableRepartitionNodeBuilder<K, V> repartitionNodeBuilder = OptimizableRepartitionNode.optimizableRepartitionNodeBuilder();
+        String sourceName = this.name;
+        StreamsGraphNode parentNode = streamsGraphNode;
 
-        final String sourceName = repartitionIfRequired(userName != null ? userName : storeBuilder.name(), repartitionNodeBuilder);
+        if (repartitionRequired) {
+            final OptimizableRepartitionNode.OptimizableRepartitionNodeBuilder<K, V> repartitionNodeBuilder = OptimizableRepartitionNode.optimizableRepartitionNodeBuilder();
+            final String repartitionTopicPrefix = userProvidedRepartitionTopicName != null ? userProvidedRepartitionTopicName : storeBuilder.name();
+            sourceName = createRepartitionSource(repartitionTopicPrefix, repartitionNodeBuilder);
 
-        StreamsGraphNode parentNode = streamsGraphNode;
+            if (repartitionNode == null || userProvidedRepartitionTopicName == null) {
```

Comment:
On the first invocation, the `repartitionNode` is always null, so we'll build it and set the variable.  But on subsequent calls, only if the user has provided a name for the repartition topic will we re-use the repartition graph node.

This maintains compatibility as users not naming the repartition topic will get the same topology and re-using  the `KGroupedStream` results in an `InvalidToplogyException` when building the topology

Reply:
> and re-using the `KGroupedStream` results in an `InvalidToplogyException` when building the topology

I thought, if there is no user topic-name, old code would create multiple repartition topics? And re-using `KGroupedStream` only throughs if there is a user topic-name (and this restriction is lifted with this PR)

Reply:
I had a hard time reasoning about the or condition and what exactly happens when there is a name / isn't a name / repartition isn't required at all.

Can we consider breaking these up into three separate cases to make the code more linear?

Reply:
>I thought, if there is no user topic-name, old code would create multiple repartition topics? And re-using `KGroupedStream` only throughs if there is a user topic-name (and this restriction is lifted with this PR)

That's correct.  I may not have been clear above, but what I  meant is that this change won't break compatibility with users currently **_not providing_** a repartition topic name as it will create multple reparitiion topics.  Does that make sense?

Reply:
>I thought, if there is no user topic-name, old code would create multiple repartition topics? And re-using `KGroupedStream` only throughs if there is a user topic-name (and this restriction is lifted with this PR)

That's correct.  I may not have been clear above, but what I  meant is that this change won't break compatibility with users currently **_not providing_** a repartition topic name as it will create multiple repartition topics.  Does that make sense?

Reply:
>Can we consider breaking these up into three separate cases to make the code more linear?

Repartitons are required when the `repartitionRequired` flag is true as passed in from the parent `KStream` instance https://github.com/apache/kafka/pull/6265/files#diff-2585c8864b9121454af6d88adba60975R81.  A name for the repartition topic is always required; it's just a matter of using a generated one or the user provided one.

So we are left with just two cases:

1. On the first call the to `GroupedStreamAggregateBuilder#build()` method the `reparitionNode` will always be `null`, so we need to create it, regardless if the user has provided a reparation topic name or not.
2. The other case is when we make subsequent `GroupedStreamAggregateBuilder#build()` calls to the same `GroupedStreamAggregateBuilder` instance; then the name does come into play.  If the user **_has provided_** a repartition topic name, we would reuse the same repartition node. Otherwise, it's null, so we know to create a new repartition node. 

I originally had this:
```java
if (repartitionNode == null) {
    repartitionNode = repartitionNodeBuilder.build();
} else if (userProvidedRepartitionTopicName == null) {
    repartitionNode = repartitionNodeBuilder.build();
}

But I felt that having multiple lines with the same variable assignment did not make sense.  What do think about me adding my explanation here to line  86 as comments?

Reply:
Yeah, comments would help. I think the logic is correct, but it's just a little complex to reason about. Some stuff I came to realize only by poking around the code:
1. there is a 1:1 relationship between a GroupedStreamAggregateBuilder instance and a KGroupedStreamImpl instance (i.e., we construct the builder in the constructor of the grouped stream node), so when we cache a repartition node in the builder, it's exactly like caching it in the KGroupedStream node, which is intuitively what we wanted to do.
2. If repartition is not required, we don't build a repartition node at all, but just bypass it in the graph.
3. Although we cache the repartition node in an instance field, we actually only use it if the node is user-named. If the name is auto-generated, we overwrite the repartition node every time before adding it to the graph.

2 and 3 were basically what I had in mind as separate branches. Something like:
* if repartition not required, just connect the agg node directly to the source node
* if repartition is required and not user-named, make a brand-new repartition node, and hook it up to the source node and hook the agg node up to the repartition node
* if the repartition is required and user-named, then check if there's already a repartition node (if not, create one, hook it up to the source node, and cache it), then hook the agg node up to the repartition node

This is just to clarify my comment above. If you prefer to clarify the flow with the explanation you gave, that's fine as well.

Reply:
> That's correct. I may not have been clear above, but what I meant is that this change won't break compatibility with users currently **_not providing_** a repartition topic name as it will create multiple repartition topics thus keep their topology the same. Does that make sense?

Ack. Just wanted to make sure we are on the same page :)

Reply:
>2 and 3 were basically what I had in mind as separate branches.

We already have this 

1.  Line 84 is the guard for checking if a repartition is not required, if not then we do skip all the repartition node creation steps and just add the agg node to the source node.
2. If a repartition is required we drop into the block and make the necessary node creation and graph links.

I'll add a comment on the boolean statement just to clarify what's going on.

Reply:
:+1: Thanks.